File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        path = []
        on_path = [False] * n

        def dfs(i):
            if i == n:
                ans.append(path.copy())
            
            for j in range(n):
                if not on_path[j]:
                    path.append(nums[j])
                    on_path[j] = True
                    dfs(i + 1)
                    path.pop()
                    on_path[j] = False
        
        dfs(0)
        return ans

File: Data Structures & Algorithms/combination-target-sum/submission-1.cpp
class Solution {
public:
    void backtrack(
        int start, 
        vector<int>& nums, 
        int target,
        vector<int>& path,
        vector<vector<int>>& ans
    ) {
        if (target < 0) {
            return;
        }

        if (target == 0) {
            ans.push_back(path);
            return;
        }

        for (int i = start; i < nums.size(); i++) {
            if (target < nums[i]) {
                break;
            }

            path.push_back(nums[i]);
            backtrack(i, nums, target - nums[i], path, ans);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<int> path;
        vector<vector<int>> ans;
        backtrack(0, nums, target, path, ans);

        return ans;
    }
};
